File: lambdas/src/test/kotlin/scorcerer/utils/BracketScoringTest.kt
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Test
import scorcerer.server.db.tables.MatchResult
import scorcerer.server.db.tables.MatchRound
import scorcerer.utils.BracketScoring
import scorcerer.utils.BracketScoring.RunMatch

/**
 * These scenarios are mirrored exactly in the frontend
 * `app/util/__tests__/bracket-scoring.test.ts`. If you change a fixture or an
 * expected total here, change it there too — the two scorers must agree.
 */
internal class BracketScoringTest {
    private val home = MatchResult.HOME
    private val away = MatchResult.AWAY
    private val r32 = MatchRound.ROUND_OF_THIRTY_TWO
    private val r16 = MatchRound.ROUND_OF_SIXTEEN

    @Test
    fun firstCorrectPickEarnsBaseOnlyStreakBonusZero() {
        val result = BracketScoring.scoreRun(listOf(RunMatch(r32, home, home)))
        result.totalPoints shouldBe 1
        result.perMatch[0] shouldBe BracketScoring.MatchScore(1, 0, true)
        result.currentStreak shouldBe 1
    }

    @Test
    fun consecutiveCorrectPicksEarnEscalatingStreakBonusCappedAtThree() {
        // 5 correct R32: bases 5, bonuses 0+1+2+3+3=9 -> 14
        val result = BracketScoring.scoreRun((1..5).map { RunMatch(r32, home, home) })
        result.totalPoints shouldBe 14
        result.currentStreak shouldBe 5
    }

    @Test
    fun roundBasesEscalateThroughTheRoundsWithStreakBonus() {
        val run = listOf(
            RunMatch(MatchRound.ROUND_OF_THIRTY_TWO, home, home),
            RunMatch(MatchRound.ROUND_OF_SIXTEEN, home, home),
            RunMatch(MatchRound.QUARTER_FINAL, home, home),
            RunMatch(MatchRound.SEMI_FINAL, home, home),
            RunMatch(MatchRound.FINAL, home, home),
        )
        // bases 1+3+5+8+12=29, bonuses 0+1+2+3+3=9 -> 38
        BracketScoring.scoreRun(run).totalPoints shouldBe 38
    }

    @Test
    fun thirdPlacePlayoffScoresLikeASemiFinal() {
        val run = listOf(
            RunMatch(MatchRound.SEMI_FINAL, home, home),
            RunMatch(MatchRound.THIRD_PLACE_PLAYOFF, home, home),
            RunMatch(MatchRound.FINAL, home, home),
        )
        // bases 8+8+12=28, bonuses 0+1+2=3 -> 31
        val result = BracketScoring.scoreRun(run)
        result.totalPoints shouldBe 31
        result.perMatch[1] shouldBe BracketScoring.MatchScore(8, 1, true)
    }

    @Test
    fun aMissResetsTheStreak() {
        val run = listOf(
            RunMatch(r32, home, home),
            RunMatch(r32, home, home),
            RunMatch(r32, home, away),
            RunMatch(r16, away, away),
        )
        // (1+0) + (1+1) + 0 + (3+0) = 6
        val result = BracketScoring.scoreRun(run)
        result.totalPoints shouldBe 6
        result.currentStreak shouldBe 1
    }

    @Test
    fun unpredictedMatchCountsAsAMiss() {
        val run = listOf(
            RunMatch(r32, home, home),
            RunMatch(r32, null, away),
            RunMatch(r32, away, away),
        )
        // (1+0) + 0 + (1+0) = 2
        BracketScoring.scoreRun(run).totalPoints shouldBe 2
    }

    @Test
    fun pendingMatchesNeitherScoreNorBreakTheStreak() {
        val run = listOf(
            RunMatch(r32, home, home),
            RunMatch(MatchRound.SEMI_FINAL, home, null),
            RunMatch(r32, home, home),
        )
        // (1+0) + pending + (1+1) = 3; streak continues through pending
        val result = BracketScoring.scoreRun(run)
        result.totalPoints shouldBe 3
        result.currentStreak shouldBe 2
        result.perMatch[1].correct shouldBe null
    }
}

File: frontend/app/util/__tests__/bracket-scoring.test.ts
import { RunMatch, scoreRun } from "@/app/util/bracket-scoring"

/**
 * These scenarios are mirrored exactly in the backend
 * `lambdas/.../scorcerer/utils/BracketScoringTest.kt`. If you change a fixture or
 * an expected total here, change it there too — the two scorers must agree.
 */

const r32 = "ROUND_OF_THIRTY_TWO" as const
const r16 = "ROUND_OF_SIXTEEN" as const
const home = "HOME" as const
const away = "AWAY" as const

describe("bracket-scoring", () => {
    it("first correct pick earns base only (streak bonus = 0)", () => {
        const result = scoreRun([{ round: r32, pick: home, actual: home }])
        expect(result.totalPoints).toBe(1)
        expect(result.perMatch[0]).toEqual({ basePoints: 1, bonusPoints: 0, correct: true })
        expect(result.currentStreak).toBe(1)
    })

    it("consecutive correct picks earn escalating streak bonus, capped at +3", () => {
        // 5 correct R32 picks: bases 5, bonuses 0+1+2+3+3=9 -> 14
        const run: RunMatch[] = Array.from({ length: 5 }, () => ({ round: r32, pick: home, actual: home }))
        const result = scoreRun(run)
        expect(result.totalPoints).toBe(14)
        expect(result.currentStreak).toBe(5)
    })

    it("escalates base points through the rounds with streak bonus", () => {
        const run: RunMatch[] = [
            { round: "ROUND_OF_THIRTY_TWO", pick: home, actual: home },
            { round: "ROUND_OF_SIXTEEN", pick: home, actual: home },
            { round: "QUARTER_FINAL", pick: home, actual: home },
            { round: "SEMI_FINAL", pick: home, actual: home },
            { round: "FINAL", pick: home, actual: home },
        ]
        // bases 1+3+5+8+12=29, bonuses 0+1+2+3+3=9 -> 38
        expect(scoreRun(run).totalPoints).toBe(38)
    })

    it("scores the third-place playoff like a semi-final", () => {
        const run: RunMatch[] = [
            { round: "SEMI_FINAL", pick: home, actual: home },
            { round: "THIRD_PLACE_PLAYOFF", pick: home, actual: home },
            { round: "FINAL", pick: home, actual: home },
        ]
        // bases 8+8+12=28, bonuses 0+1+2=3 -> 31
        const result = scoreRun(run)
        expect(result.totalPoints).toBe(31)
        expect(result.perMatch[1]).toEqual({ basePoints: 8, bonusPoints: 1, correct: true })
    })

    it("resets the streak on a miss", () => {
        const run: RunMatch[] = [
            { round: r32, pick: home, actual: home },
            { round: r32, pick: home, actual: home },
            { round: r32, pick: home, actual: away },
            { round: r16, pick: away, actual: away },
        ]
        // (1+0) + (1+1) + 0 + (3+0) = 6
        const result = scoreRun(run)
        expect(result.totalPoints).toBe(6)
        expect(result.currentStreak).toBe(1)
    })

    it("treats an unpredicted match as a miss", () => {
        const run: RunMatch[] = [
            { round: r32, pick: home, actual: home },
            { round: r32, pick: undefined, actual: away },
            { round: r32, pick: away, actual: away },
        ]
        // (1+0) + 0 + (1+0) = 2
        expect(scoreRun(run).totalPoints).toBe(2)
    })

    it("does not score or break the streak on pending matches", () => {
        const run: RunMatch[] = [
            { round: r32, pick: home, actual: home },
            { round: "SEMI_FINAL", pick: home, actual: undefined },
            { round: r32, pick: home, actual: home },
        ]
        // (1+0) + pending + (1+1) = 3; streak continues through pending
        const result = scoreRun(run)
        expect(result.totalPoints).toBe(3)
        expect(result.currentStreak).toBe(2)
        expect(result.perMatch[1].correct).toBeUndefined()
    })
})
